ksqldb-rest-app/src/main/java/io/confluent/ksql/rest/server/execution/ListSourceExecutor.java

```diff
@@ -208,13 +238,86 @@ private static SourceDescriptionWithWarnings describeSource(
+    final Set<String> allTopics = new HashSet<>();
+    // Get topics and offsets per running query
+    for (RunningQuery query : sinkQueries) {
+      final QueryId queryId = query.getId();
+      final String applicationId =
+          QueryApplicationId.build(ksqlConfig, true, queryId);
+      final Map<TopicPartition, OffsetAndMetadata> topicAndConsumerOffsets =
+          serviceContext.getConsumerGroupClient().listConsumerGroupOffsets(applicationId);
+      offsetsPerQuery.put(applicationId, topicAndConsumerOffsets);
+      final Set<String> topics = topicAndConsumerOffsets.keySet().stream()
+          .map(TopicPartition::topic)
+          .collect(Collectors.toSet());
+      topicsPerQuery.put(applicationId, topics);
+      allTopics.addAll(topics);
+    }
+    // Get topics descriptions and start/end offsets
+    final Map<String, TopicDescription> sourceTopicDescriptions =
+        serviceContext.getTopicClient().describeTopics(allTopics);
+    final Map<TopicPartition, Long> topicAndStartOffsets =
+        serviceContext.getTopicClient().listTopicsStartOffsets(allTopics);
+    final Map<TopicPartition, Long> topicAndEndOffsets =
+        serviceContext.getTopicClient().listTopicsEndOffsets(allTopics);
+    // Build consumer offsets summary
+    for (Entry<String, Set<String>> entry : topicsPerQuery.entrySet()) {
+      for (String topic : entry.getValue()) {
+        sourceConsumerOffsets.add(
+            new QueryOffsetSummary(
+                entry.getKey(),
+                topic,
+                consumerPartitionOffsets(
+                    sourceTopicDescriptions.get(topic),
+                    topicAndStartOffsets,
+                    topicAndEndOffsets,
+                    offsetsPerQuery.get(entry.getKey()))));
+      }
+    }
+    return sourceConsumerOffsets;
+  }
+
+  private static List<ConsumerPartitionOffsets> consumerPartitionOffsets(
+      final TopicDescription topicDescription,
+      final Map<TopicPartition, Long> topicAndStartOffsets,
+      final Map<TopicPartition, Long> topicAndEndOffsets,
+      final Map<TopicPartition, OffsetAndMetadata> topicAndConsumerOffsets
+  ) {
+    final List<ConsumerPartitionOffsets> consumerPartitionOffsets = new ArrayList<>();
+    for (TopicPartitionInfo topicPartitionInfo : topicDescription.partitions()) {
+      final TopicPartition tp = new TopicPartition(topicDescription.name(),
+          topicPartitionInfo.partition());
+      final Long startOffsetResultInfo = topicAndStartOffsets.get(tp);
+      final Long endOffsetResultInfo = topicAndEndOffsets.get(tp);
+      final OffsetAndMetadata offsetAndMetadata = topicAndConsumerOffsets.get(tp);
+      consumerPartitionOffsets.add(
+          new ConsumerPartitionOffsets(
+              topicPartitionInfo.partition(),
+              startOffsetResultInfo,
+              endOffsetResultInfo,
+              offsetAndMetadata != null ? offsetAndMetadata.offset() : 0
```

Comment:
Under what situations would this be null?

Reply:
I've seen this when consumers haven't poll from a topic-partition yet, the list of offset and metadata only returns the ones that exist at the moment. Adding a short comment to clarify this.